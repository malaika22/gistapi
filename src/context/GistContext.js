import { createContext, useEffect, useState } from 'react';
import { useFetch } from '../hooks/useFetch';
import { getPublicGists } from '../services/gistService';
import PropTypes from 'prop-types';

export const GistContext = createContext();

export const GistProvider = ({ children }) => {
  const { data, loading: fetchGistsLoading } = useFetch(getPublicGists);
  const [gistList, setGistList] = useState([]);

  useEffect(() => {
    setGistList(data);
  }, [data]);

  return (
    <GistContext.Provider value={{ gistList, setGistList, fetchGistsLoading }}>
      {children}
    </GistContext.Provider>
  );
};

GistProvider.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]).isRequired
};
