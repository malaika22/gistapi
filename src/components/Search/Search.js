import React, { useContext, useEffect, useState } from 'react';
import styled from 'styled-components';
import Octicon from 'react-octicon';
import { useDebounce } from '../../hooks/useDebounce';
import { useQueryFetch } from '../../hooks/useQueryFetch';
import { getGistForUser } from '../../services/gistService';
import { GistContext } from '../../context/GistContext';

export const Search = () => {
  const [searchValue, setSearchedValue] = useState('');
  const searchQuery = useDebounce(searchValue, 2000);
  const { setGistList } = useContext(GistContext);
  const { data: gistsData } = useQueryFetch(getGistForUser, searchQuery);

  const onGistSearch = ({ target: { value } }) => {
    setSearchedValue(value);
  };

  useEffect(() => {
    setGistList(gistsData);
  }, [gistsData]);

  return (
    <Wrapper>
      <InputBox>
        <Octicon name="search" />
        <Input
          value={searchValue}
          onChange={onGistSearch}
          placeholder="Search Gists for the username"
        />
      </InputBox>
    </Wrapper>
  );
};

const Wrapper = styled.div`
  padding: 8px;
  background-color: #ffffff;
  font-size: 14px;
  line-height: 1.5;
  border-radius: 6px;
  margin: 0 16px;
`;

const InputBox = styled.div`
  border-radius: 4px;
  display: flex;
  width: 400px;
`;

const Input = styled.input`
  border: none;
  width: 100%;
  font-size: 16px;

  &:focus {
    outline: 0;
  }
`;

export default Search;
