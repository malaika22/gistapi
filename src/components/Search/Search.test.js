import { render } from '@testing-library/react';
import Search from './Search';
import { GistProvider } from 'context/GistContext';

describe('Search', () => {
  it('should render successfully', () => {
    const { baseElement } = render(
      <GistProvider>
        <Search />
      </GistProvider>
    );
    expect(baseElement).toBeTruthy();
  });
  it('should match snapshot', () => {
    const { asFragment } = render(
      <GistProvider>
        <Search />
      </GistProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
