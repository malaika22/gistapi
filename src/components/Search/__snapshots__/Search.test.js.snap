// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bgqQcB ighHLN"
  >
    <div
      class="sc-gTRrQi cLuoSN"
    >
      <span
        class="octicon octicon-search"
      />
      <input
        class="sc-ewnqHT eQrKCi"
        placeholder="Search Gists for the username"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
